developer_docs/how_to_create_a_plugin.asciidoc

```diff
@@ -0,0 +1,2156 @@
+    extend_model PuppetHostExtensions
+    add_helper PuppetFacetHelper
+    add_tabs :puppet_tabs
+    api_view :list => 'api/v2/puppet_facets/base', :single => 'api/v2/puppet_facets/single_host_view'
+    template_compatibility_properties :environment_id, :puppet_proxy_id, :puppet_ca_proxy_id
+    set_dependent_action :destroy # requires #21657, Foreman >= 1.19
+  end
+end
+----
+This is being re-worked into a proper plugin API via #13417, it's highly
+recommended to use that when available and not use internal APIs.
+
+[[facets.register-method]]
+==== Facets.register method
+
+this method takes two parameters and an initialization block:
+
+* *facet_model* A class that will be used as a model.
+* *facet_name* (optional) a new name for the relation in the host model.
+
+The initialization block exposes the following DSL:
+
+[[extend_model]]
+==== #extend_model
+
+* *extension_module* Module to be included in the host model
+
+Use this extension point if you want to add functionality to the
+Host::Managed object. Be aware that not every host will contain a valid
+instance of your facet.
+
+[[add_helper]]
+==== #add_helper
+
+* *facet_helper* Helper module to be included in host's view.
+
+Use this extension point to add methods that will be available to the
+View phase. You will be able to use those methods in your facet's
+related templates.
+
+[[add_tabs]]
+==== #add_tabs
+
+* *tabs* The parameter can be either a hash or a symbol that points to a
+method in helper.
+
+In addition to the main facet's tab (that is declared by
+`app/views/my_facets/_my_facet.html.erb`) each facet can declare
+additional tabs to be shown in the UI. The declaration can be either
+static - a static hash of keys and tab templates, or dynamic - the hash
+will be generated for each host.
+
+The hash should contain the following information:
+
+* *key* should be an identifier that will be used by the UI framework to
+identify the new tab
+
+* *value* should be a value that will be passed to _render_ method - it
+can be a string representing a template or an object. The _render_ call
+will set `f` parameter to the value of host's form, if you want to add
+parameters to be passed at the submit method. +
+Example:
+
+[source, ruby]
+----
+tabs_hash = {
+  :puppetclasses => 'puppet_facets/puppetclasses_tab', #will call puppetclasses_tab.html.erb template
+  :facet_tab_example => SomeModel.first, #will try to match a template for SomeModel.
+}
+----
+*static declaration*
+
+[source, ruby]
+----
+Rails.application.config.to_prepare do
+  Facets.register(PuppetFacet) do
+    tabs_hash = {
+      :puppetclasses => 'puppet_facets/puppetclasses_tab', #will call puppetclasses_tab.html.erb template
+      :facet_tab_example => SomeModel.first, #will try to match a template for SomeModel.
+    }
+
+    add_tabs tabs_hash #will generate two more tabs for each host.
+  end
+end
+----
+*dynamic declaration*
+
+In `my_facet_helper.rb`:
+
+.my_facet_helper.rb
```

Comment:
I have just noticed, do we need the file name above the code examples twice in quick succession?
![file-names](https://user-images.githubusercontent.com/2664090/56879168-be5d5f00-6a57-11e9-8513-efbc50080114.png)

Reply:
Removed the manual line. Now I am using only asciidoc's filename notation.